Mathlib/Tactic/Trans.lean

```diff
@@ -0,0 +1,111 @@
+import Lean
+import Mathlib.Tactic.Symm
+
+
+namespace Mathlib.Tactic.Trans
+open Lean Meta Elab
+open Mathlib.Tactic.Symm
+
+/- Environment extension storing transitivity lemmas -/
+initialize transExtension : SimpleScopedEnvExtension (Name × Array DiscrTree.Key) (DiscrTree Name) ←
+  registerSimpleScopedEnvExtension {
+    name := `trans
+    addEntry := fun dt (n, ks) => dt.insertCore ks n
+    initial := {}
+  }
+
+/- checks whether there is some `y` so that `xyHyp` has type `rel x y` and `yzHyp` has type `rel y z`.
+-/
+def areTransHyps(rel x z xyHyp yzHyp : Expr) : MetaM Bool :=
+  try
+    let y ← mkFreshExprMVar (some <| ← inferType x)
+    let modelPair ← mkAppM ``PProd.mk #[mkAppN rel #[x, y], mkAppN rel #[y, z]]
+    let hypPair ← mkAppM ``PProd.mk #[xyHyp, yzHyp]
+    isDefEq modelPair hypPair
+    -- pure true
+  catch _ => pure false
+
+/-- add transitivity attribute if valid-/
+def transAttr : AttributeImpl where
+  name := `trans
+  descr := "transitive relation"
+  add decl _ kind := do
+    MetaM.run' do
+      let declTy := (← getConstInfo decl).type
+      let (xs, _, targetTy) ← withReducible <| forallMetaTelescopeReducing declTy
+      if xs.size < 2 then
+        throwError "@[trans] attribute only applies to lemmas proving x ∼ y → y ∼ z → z ∼ x, got {declTy} with too few arguments"
+      else
+        match ← relationAppM? targetTy with
+        | some (rel, x, z) =>
+          let xyHyp ← inferType xs[xs.size -2]!
+          let yzHyp ← inferType xs[xs.size -1]!
+          let y ← mkFreshExprMVar (some <| ← inferType x)
+          let modelPair ← mkAppM ``PProd.mk #[← mkAppM' rel #[x, y], ← mkAppM' rel #[y, z]]
+          let hypPair ← mkAppM ``PProd.mk #[xyHyp, yzHyp]
+          unless (← isDefEq hypPair modelPair) do
+            throwError "@[trans] attribute only applies to lemmas proving x ∼ y → y ∼ z → z ∼ x, got {declTy}"
+          let key ← withReducible <| DiscrTree.mkPath rel
+          transExtension.add (decl, key) kind
+        | none =>
+          throwError "@[trans] attribute only applies to lemmas proving x ∼ y → y ∼ z → z ∼ x, got {declTy}"
+
+initialize registerBuiltinAttribute transAttr
+
+/-- look up transitivity lemmas -/
+def transLemmas (env : Environment) : DiscrTree Name :=
+  transExtension.getState env
+
+/-- transitivity lemmas from the typeclass -/
+theorem simpleTrans {rel : α → α → Prop}{a b c : α}[Trans rel rel rel] :
+    rel a b → rel b c → rel a c  := trans
+
+open Lean.Elab.Tactic
+/-- transitivity tactic implementation -/
```

Comment:
This should be the doc string for the trans tactic itself. You can copy it from lean 3.

Reply:
Done.